packages/transform/src/lib/TransformMiddleware.ts

```diff
@@ -0,0 +1,485 @@
+
+    await this.updateMetadataPath(key, this.objectPathKeys(data));
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.SetMany](payload: Payload.SetMany): Promise<Payload.SetMany> {
+    const { entries } = payload;
+    const { before } = this.context;
+
+    payload.entries = entries.map(({ key, path, value }) => ({ key, path, value: before(value as BeforeValue, key, path ?? null) }));
+    await this.provider[Method.SetMany](payload);
+
+    entries.forEach(async ({ key, path }) => {
+      const { data } = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path });
+
+      await this.updateMetadataPath(key, this.objectPathKeys(data));
+    });
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Update]<StoredValue = AfterValue, ReturnValue = BeforeValue>(
+    payload: Payload.Update<StoredValue, ReturnValue>
+  ): Promise<Payload.Update<StoredValue, ReturnValue>> {
+    const { key, hook } = payload;
+    const { before } = this.context;
+    const getBefore = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+    payload.hook = (value: StoredValue) => before(hook!(value, key) as BeforeValue, key, null) as Awaitable<ReturnValue>;
+
+    await this.provider[Method.Update](payload as unknown as Payload.Update<AfterValue, ReturnValue>);
+
+    const getAfter = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+    if (this.getChangedKeys(getBefore.data, getAfter.data).length! > 0) {
+      await this.updateMetadataPath(key, this.getChangedKeys(getBefore.data, getAfter.data));
+    }
+
+    return payload;
+  }
+
+  @PostProvider()
+  public async [Method.Dec](payload: Payload.Dec): Promise<Payload.Dec> {
+    const { after, before } = this.context;
+    const { key } = payload;
+    const getBefore = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+    const beforeValue = after(getBefore.data!, key, null);
+
+    await this.provider[Method.Set]({ method: Method.Set, errors: [], key, path: [], value: beforeValue });
+    await this.provider[Method.Dec](payload);
+
+    const getAfter = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+    const afterValue = before(getAfter.data as BeforeValue, key, null);
+
+    await this.provider[Method.Set]({ method: Method.Set, errors: [], key, path: [], value: afterValue });
+
+    if (this.getChangedKeys(getBefore.data, getAfter.data).length) {
+      await this.updateMetadataPath(key, this.getChangedKeys(getBefore.data, getAfter.data));
+    }
+
+    return payload;
+  }
+
+  public async [Method.Every]<ReturnValue = BeforeValue>(payload: Payload.Every.ByHook<ReturnValue>): Promise<Payload.Every.ByHook<ReturnValue>>;
+  public async [Method.Every](payload: Payload.Every.ByValue): Promise<Payload.Every.ByValue>;
+  @PostProvider()
```

Comment:
Would this not be PreProvider with a skip metadata value?

Reply:
Yes it should be, I will take sometime to redo it to add this check.